### PR TITLE
Prepare commitment info for OsInput in prove_block

### PR DESCRIPTION
**Pull request by notlesh**
This PR calculates the commitment info for state and class tries and provides them to the `OsInput`.

Issue Number: N/A

## Type

- [ ] feature
- [ ] bugfix
- [ ] dev (no functional changes, no API changes)
- [ ] fmt (formatting, renaming)
- [ ] build
- [ ] docs
- [ ] testing

## Breaking changes?

- [ ] yes
- [ ] no

### CODE REVIEW

File: crates/bin/prove_block/src/rpc_utils.rs [hunk 1]
```diff
@@ -12,24 +12,36 @@ use starknet_types_core::felt::Felt;
 
 /// A `Storage` impl backed by RPC
 #[derive(Clone)]
-pub(crate) struct RpcStorage {}
+pub(crate) struct RpcStorage {
+    cache: HashMap<Vec<u8>, Vec<u8>>,
```
Comment (notlesh) [inline crates/bin/prove_block/src/rpc_utils.rs:16]: TODO: this is redundant with the `CompositeStorage` setup, needs to be removed

File: crates/bin/prove_block/src/main.rs [hunk 2]
```diff
@@ -380,10 +401,34 @@ async fn main() -> Result<(), Box<dyn Error>> {
     for (contract_address, _storage_proof) in storage_proofs {
         let previous_storage_proof =
             previous_storage_proofs.get(&contract_address).expect("failed to find previous storage proof");
+        let contract_storage_root = previous_storage_proof.contract_data.as_ref().unwrap().root.into();
```
Comment (HermanObst) [inline crates/bin/prove_block/src/main.rs:404]: @notlesh call this `previous_contract_storage_proof`

File: crates/bin/prove_block/src/main.rs [hunk 3]
```diff
@@ -380,10 +401,34 @@ async fn main() -> Result<(), Box<dyn Error>> {
...
+        let previous_storage_entries = previous_storage_changes_by_contract
```
Comment (HermanObst) [inline crates/bin/prove_block/src/main.rs:405]: This also should be called `previous_storage_changes_entries`

File: crates/bin/prove_block/src/rpc_utils.rs (hunk 1 above)
Comment (odesenfans) [reply crates/bin/prove_block/src/rpc_utils.rs:16]: Yes, I don't see the point of this?

File: crates/bin/prove_block/src/rpc_utils.rs [hunk 4]
```diff
@@ -139,27 +153,29 @@ pub(crate) async fn get_storage_proofs(
...
             merge_chunked_storage_proofs(chunked_storage_proofs)
         };
 
-        storage_proofs.insert(contract_address, storage_proof);
+        assert!(!storage_proofs.contains_key(contract_address), "Can't overwrite storage proof!");
+        storage_proofs.insert(*contract_address, storage_proof);
     }
 
-    Ok(storage_proofs)
+    Ok((storage_proofs, storage_changes_by_contract))
```
Comment (odesenfans) [inline crates/bin/prove_block/src/rpc_utils.rs:178]: I don't understand what you're doing with these modifications.

File: crates/bin/prove_block/src/main.rs [hunk 5]
```diff
@@ -345,6 +347,25 @@ async fn main() -> Result<(), Box<dyn Error>> {
...
+                        }
+                        TrieNode::Edge { child, path } => {
+                            // log::info!("writing edge node...");
+                            let fact = EdgeNodeFact::new((*child).into(), NodePath(path.value.to_biguint()), Length(path.len))?;
+                            fact.set_fact(&mut initial_state.ffc).await?;
+                        }
+                    }
+                }
+            }
+        }
```
Comment (odesenfans) [inline crates/bin/prove_block/src/main.rs:368]: We're going to use similar code for the global state and class tries as well, make it a function.

File: crates/bin/prove_block/src/main.rs [hunk 6]
```diff
@@ -489,51 +534,34 @@ async fn main() -> Result<(), Box<dyn Error>> {
...
-    //         &contract_indices,
-    //         &mut initial_state.ffc.clone_with_different_hash::<PoseidonHash>(),
-    //     )
-    //     .await
-    //     .unwrap_or_else(|e| panic!("Could not create contract class commitment info: {:?}", e));
+    let contract_indices: HashSet<TreeIndex> =
+        contract_states.keys().chain(contract_storages.keys()).map(|address| address.to_biguint()).collect();
+    let contract_indices: Vec<TreeIndex> = contract_indices.into_iter().collect();
+
+    let final_state = initial_state.clone().apply_commitment_state_diff(blockifier_state.to_state_diff()).await?;
```
Comment (odesenfans) [inline crates/bin/prove_block/src/main.rs:542]: We should create the commitment info from RPC data instead of reapplying the changes from Blockifier.

**Review by odesenfans — CHANGES_REQUESTED** [review]
This should be improved a little bit before merging. + the CI is not passing yet.

File: crates/bin/prove_block/src/main.rs [hunk 7]
```diff
@@ -2,6 +2,7 @@ use std::collections::{HashMap, HashSet};
 use std::error::Error;
 
 use blockifier::state::cached_state::{CachedState, GlobalContractCache};
+use blockifier::state::state_api::State as _;
```
[the remark this answers is not in the document]
Comment (notlesh) [reply crates/bin/prove_block/src/main.rs:5]: It's used when you want a trait in scope but do not need to use it directly. I always thought it added a little bit of clarity as to why it's being used, but I don't have a strong opinion about it...

**Review by odesenfans — APPROVED** [review]

**Pull request merged**